google-cloud-logging/src/main/java/com/google/cloud/logging/BatchingWriter.java

```diff
@@ -85,43 +81,50 @@ synchronized void stopJob() {
     flushJob.cancel(false);
   }
 
-  void add(com.google.logging.v2.LogEntry entry) throws InterruptedException {
-    pending.acquire(1);
-    synchronized (currentBatch) {
-      currentBatch.add(entry);
-      if (currentBatch.size() == batchSize) {
-        final WriteLogEntriesRequest request =
-            requestPrototype.toBuilder().addAllEntries(currentBatch).build();
-        currentBatch.clear();
+  void add(com.google.logging.v2.LogEntry entry) throws FlowController.FlowControlException {
+    long entrySize = entry.getSerializedSize();
 
+    flowController.reserve(1, entrySize);
+    synchronized (accumulator) {
+      accumulator.add(entry, entrySize, null);
+      while (accumulator.hasBatch()) {
+        final WriteLogEntriesRequest request =
+            requestPrototype.toBuilder().addAllEntries(accumulator.batch()).build();
+        final long size = accumulator.bytes();
         // Whoever calls send serializes the proto; so we do it off-thread.
         // This gives better CPU utilization if there are few producer threads
         // on a many-core machine.
+        //
+        // BUG(pongad): send() puts future in the map too late. To make flush work properly,
+        // we must guarantee adding future to map before this method returns.
         executor.execute(
             new Runnable() {
               @Override
               public void run() {
-                send(request);
+                send(request, size);
```

Comment:
I think it might be better to remove the off thread execution (at least for the time being). It should solve the flush race condition. I'm also not convinced that CPU utilization of serializing protos will be a bottleneck here and worth the complexity

Reply:
I buy the argument that we should `send` in-thread right now to simplify things. That said, when I desktop-tested this code a while back, one thread writing trivial logs pushed my machine to 400% CPU so I think the bottleneck could be real.

I'll add a TODO for off-thread serialization instead.